docs/github-app.adoc

```diff
@@ -87,7 +87,8 @@ Fill out the form:
 - App ID: the github app ID, it can be found in the 'About' section of your GitHub app in the general tab.
 - API endpoint (optional, only required for GitHub enterprise this will only show up if a GitHub enterprise server is configured).
 - Key: click add, paste the contents of the converted private key
-- Passphrase: do not fill this field, it will be ignored
+- Advanced: (optional) If you've installed your same GitHub app on multiple organisations you need the next step
+  * Owner: the name of the organisation or user, i.e. jenkinsci for https://github.com/jenkinsci
```

Comment:
So under this approach you cannot reuse the Jenkins credentials object across orgs, and you need to duplicate the private key. Not as smooth as having Jenkins figure out which installation to work with, but better than nothing, right?

Reply:
yeah I couldn't see a way to do that, as we don't have the context from the build job at all

Reply:
Right, it would be a more intrusive patch since you would need to do something along the lines of defining a thread-local org and setting this from a new overload of `Connector.connect` which passes a `repoOwner` and then switching as many callers as possible to the new overload. Not sure it is worth the effort since I presume the common case is for the Jenkins installation to be used on only a single org.

Reply:
>Not sure it is worth the effort since I presume the common case is for the Jenkins installation to be used on only a single org.

What about for GitHub Enterprise? It doesn't seem to be possible to install an app at the Enterprise level. My Jenkins instance works on repos across many organizations in a GitHub Enterprise instance.

Reply:
I do not have GHE to test against but in that case with the current patch you would need to first install the app in each such org, then create a separate Jenkins credentials item for each installation.

Reply:
At my company we've been using a GitHub App on GHE to provide credentials for the last six months or so. Unfortunately we weren't able to open source the plugin, but we were very happy to see this feature becoming available in the branch source plugin.

> Not sure it is worth the effort since I presume the common case is for the Jenkins installation to be used on only a single org.

For our use case we need to support for multiple orgs as we have a single GitHub App which gets installed in every org, this would probably be a common approach for others who want to use this with GHE.

Looking at the code, it seems very similar to the approach we took so I don't expect you'll run into any issues relating to GHE. I'll try to find some time to build a new version of this plugin to test with our GHE instance and we can help fix any bugs we may run into.

Reply:
the version that's on master has been tested on GHE, some of the companies using it in production have GHE.

in regards to multiple orgs with the same app it could be pretty straightforward to define the credentials with jcasc:

```
credentials:
  system:
    domainCredentials:
      - credentials:
        - gitHubApp:
            appID: "1111"
            description: "GitHub app team 1"
            id: "github-app-team-1"
            # apiUri: https://my-custom-github-enterprise.com/api/v3 # optional only required for GitHub enterprise
            privateKey: "${GITHUB_APP_KEY}"
			owner: team1
        - gitHubApp:
            appID: "1111"
            description: "GitHub app team 2"
            id: "github-app-team-2"
            # apiUri: https://my-custom-github-enterprise.com/api/v3 # optional only required for GitHub enterprise
            privateKey: "${GITHUB_APP_KEY}"
			owner: team2
        - gitHubApp:
            appID: "1111"
            description: "GitHub app team 3"
            id: "github-app-team-3"
            # apiUri: https://my-custom-github-enterprise.com/api/v3 # optional only required for GitHub enterprise
            privateKey: "${GITHUB_APP_KEY}"
			owner: team3
```

Reply:
Good enough IMHO

Reply:
When we have tested beta version (internal and we use GHE) I missed testing this multiple organization functionality, now we have upgraded to latest released version (2.7.1) and trying to use same app on Jobs across orgs and is not working, we need to maintain duplicate credentials. 

Maintaining same credentials for multiple orgs seems duplicate to me, is there any chance we can avoid this? I guess underlying api doesn't require the actual owner as @jglick pointed out in this comment https://github.com/jenkinsci/github-branch-source-plugin/pull/290#discussion_r401115617 

For us, since it is internal bot, it is pretty open and we don't even know who would invite this bot to their organization, and this bot does takes care of creating jenkins job through dsl scripts, which is using the standard credentials. So it is very hard to keep creating credentials on the fly for every org that this bot is being invited to. 

Thanks a bunch for all the help on maintaining this awesome plugin. Let me know your thoughts I can log a follow up JIRA.

Reply:
The

Reply:
Yes @jglick

Reply:
Not sure of the underlying limitation which let us add this owner setting, but at one point with incremental version I was able to test GitHub app without owner settings and it was working across orgs

Reply:
@nrayapari were your repos public repos across the orgs? 

If so it would likely have worked, I think the access token is just ignored or at least the scope being wrong doesn’t matter for the cloning of it

Private repos and updating the commit status should not have worked though

Reply:
Yes, those were all public repos. 

is there any way that we can get this owner based on the job config as it would already have the github repo information ?

Reply:
The credential isn’t tied to a job it’s independent, not safely, there may be some possible hacks to do it though

Reply:
My company uses Jenkins in a similar fashion as the company described by @nrayapati.

We deployed a single `jenkins` GitHub App that may be installed by any organization on our GHE instance. There are several hundred such organizations.

We allow our engineers to create jobs as needed within their own Jenkins folders and leverage the single GitHub App Credential.

Since any engineer can install our GitHub App on their organization, and any engineer can create an organization, we would be in the situation of needing to create a Credential for every organization as it's created.

I, too, would like to see support for a single global Credential to work for jobs across multiple organizations.

@nrayapati can you please file a Jira ticket (As this PR has already been merged and released) so that we can centralize the discussion about this use case on that ticket?

Reply:
https://issues.jenkins-ci.org/browse/JENKINS-62220

Reply:
Can we just use the org name in repo instead of the org name in credential config?